staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go

```diff
@@ -2038,6 +2104,65 @@ func TestAllocator(t *testing.T) {
 				deviceAllocationResult(req0SubReq1, driverA, pool1, device1, false),
 			)},
 		},
+		"tainted-two-devices": {
+			deviceTaints:     true,
+			claimsToAllocate: objects(claim(claim0, req0, classA)),
+			classes:          objects(class(classA, driverA)),
+			slices: objects(slice(slice1, node1, pool1, driverA,
+				device(device1, nil, nil).withTaints(taintNoSchedule),
+				device(device2, nil, nil).withTaints(taintNoExecute),
+			)),
+			node: node(node1, region1),
+		},
+		"tainted-one-device-two-taints": {
+			deviceTaints:     true,
+			claimsToAllocate: objects(claim(claim0, req0, classA)),
+			classes:          objects(class(classA, driverA)),
+			slices: objects(slice(slice1, node1, pool1, driverA,
+				device(device1, nil, nil).withTaints(taintNoSchedule, taintNoExecute),
+			)),
+			node: node(node1, region1),
+		},
+		"tainted-two-devices-tolerated": {
+			deviceTaints:     true,
+			claimsToAllocate: objects(claim(claim0, req0, classA).withTolerations(tolerationNoExecute)),
+			classes:          objects(class(classA, driverA)),
+			slices: objects(slice(slice1, node1, pool1, driverA,
+				device(device1, nil, nil).withTaints(taintNoSchedule),
+				device(device2, nil, nil).withTaints(taintNoExecute),
+			)),
+			node: node(node1, region1),
+			expectResults: []any{allocationResult(
+				localNodeSelector(node1),
+				deviceAllocationResult(req0, driverA, pool1, device2, false), // Only second device's taints are tolerated.
+			)},
+		},
+		"tainted-one-device-two-taints-both-tolerated": {
+			deviceTaints:     true,
+			claimsToAllocate: objects(claim(claim0, req0, classA).withTolerations(tolerationNoSchedule, tolerationNoExecute)),
+			classes:          objects(class(classA, driverA)),
+			slices: objects(slice(slice1, node1, pool1, driverA,
+				device(device1, nil, nil).withTaints(taintNoSchedule, taintNoExecute),
+			)),
+			node: node(node1, region1),
+			expectResults: []any{allocationResult(
+				localNodeSelector(node1),
+				deviceAllocationResult(req0, driverA, pool1, device1, false),
+			)},
+		},
+		"tainted-disabled": {
+			deviceTaints:     false,
+			claimsToAllocate: objects(claim(claim0, req0, classA)),
+			classes:          objects(class(classA, driverA)),
+			slices: objects(slice(slice1, node1, pool1, driverA,
+				device(device1, nil, nil).withTaints(taintNoSchedule, taintNoExecute),
+			)),
+			node: node(node1, region1),
+			expectResults: []any{allocationResult(
+				localNodeSelector(node1),
+				deviceAllocationResult(req0, driverA, pool1, device1, false),
+			)},
+		},
```

Comment:
Couple questions:
1) What is the expected behavior of taints with "All" allocation mode? Should All mode fail if a device is tainted? Probably so.
but...
2) (non-blocking) What is the expected behavior of taints with admin access? In particular, shouldn't a DaemonSet that is monitoring devices via admin access, and using All mode, schedule just fine and get access to all devices? Or should it just be sure to set the right tolerations?

Reply:
There's nothing special in both cases.

"All" mode requires that the all devices must be usable. If one is tainted, nothing gets allocated and the pod cannot get scheduled.

Admin access also doesn't bypass this check. If an admin was access to to a tainted device, they need to tolerate the taint.

Reply:
Might be worth some test cases... I'll add them.

Reply:
Done, all worked as described.